[REVIEW] Fix dask_cudf group_split behavior to handle upstream rearrange_by_divisions

rjzamora opened the pull request:
The recently-added `group_split_cudf` dispatch function does not handle `rearrange_by_divisions` correctly.  This PR adds a new test (`test_rearrange_by_divisions`), as well as a fix for the primary bug.  There are actually **two** problems with the current implementation:

1. [**Primary Bug**] The output of `group_split_cudf` should be a dictionary; with each key being a consecutive integer index, and each item being a cudf dataframe.  The current output is a list, which is clearly a bug (my mistake).

2. [Discrepancy between dask and dask_cudf] The original index values should be conserved in the `group_split_cudf` output.  In other words, each dataframe row in the output dictionary should have an index value that could be used to reconstruct the original dataframe.  To clarify the difference, here is an example of `rearrange_by_divisions` behavior with both pandas and cudf-based dask dataframes:

**Original DataFrame**:
```
    x         y
0   6  0.392580
1  19 -0.929185
2  14  0.079832
3  10 -0.159517
4   7  0.022222
5   6 -0.427793
6  18 -0.531817
7  10 -0.117476
8  10  0.222079
9   3 -0.767977
```

**Pandas-based `rearrange_by_divisions` Behavior**:
```
    x         y
9   3 -0.767977
0   6  0.392580
4   7  0.022222
5   6 -0.427793
3  10 -0.159517
7  10 -0.117476
8  10  0.222079
2  14  0.079832
1  19 -0.929185
6  18 -0.531817
```

**cudf-based `rearrange_by_divisions` Behavior** (after dictionary fix):
```
    x         y
0   3 -0.767977
0   6  0.392580
1   7  0.022222
2   6 -0.427793
0  10 -0.159517
1  10 -0.117476
2  10  0.222079
0  14  0.079832
0  19 -0.929185
1  18 -0.531817
```

Notice that the index values are not conserved for the cudf-based DataFrame result.

**Important Note**:  In the current draft of this PR, index conservation is ignored by using `check_index=False`  in `test_rearrange_by_divisions`.  Correct behavior can be accomplished by resetting the index twice (before and after `scatter_by_map`), but this seems inefficient.  Please feel free to comment on this here.

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

mrocklin commented on python/dask_cudf/dask_cudf/tests/test_core.py:
```diff
@@ -195,6 +195,29 @@ def test_set_index_w_series():
+        result = dd.shuffle.rearrange_by_divisions(
+            gdf1, "x", divisions=divisions, shuffle="tasks"
+        )
+        dd.assert_eq(expect, result, check_index=False)
```
So the second bug would be tested if we change this to `check_index=True`?

rjzamora replied:
Yes - I am currently thinking through the "best" way to address this.

mrocklin commented:
The change here seems good to me.  

What do we need to do to resolve the second issue?

rjzamora commented:
> What do we need to do to resolve the second issue?

I am trying to figure this out.  [The first commit in this PR](https://github.com/rapidsai/cudf/pull/2895/commits/ac76dfbf0e7e0e48a2192998abc7e5580695a0c0) addresses the issue, but that approach feels wrong to me.    I am looking into *lower-level* ways to avoid this pands-cudf discrepancy.

mrocklin commented:
Yeah, maybe this is something that we kick back to @harrism and team?

kkraus14 commented:
> Yeah, maybe this is something that we kick back to @harrism and team?

libcudf has no concept of index column(s) unless absolutely necessary. I'd guess it's more likely that we're just losing the index column(s) by not passing them through Cython into cudf::table constructors erroneously somewhere.

Before kicking over we should confirm we need libcudf behavior changed.

mrocklin commented:
Ah, my apologies.

rjzamora commented:
> Before kicking over we should confirm we need libcudf behavior changed.

Right - I am thinking the change(s) should be in python/cython

rjzamora commented:
**Update**:  I modified the `scatter_by_map` DataFrame API and the `scatter_to_frames` libcudf wrapper to optionally preserve the index of each row after the call to `cpp_scatter_to_tables`.  I also updated `test_dataframe_scatter_by_map` and `test_rearrange_by_divisions` -- Feel free to suggest alternative solutions.

harrism commented:
Is this still a 0.10 PR? It is currently an orphan.

kkraus14 commented:
Let's push to 0.11.

kkraus14 commented on python/cudf/cudf/_lib/copying.pyx:
```diff
@@ -236,14 +236,15 @@ def copy_range(out_col, in_col, int out_begin, int out_end,
     ----------
     source : Column or list of Columns
     maps : non-null column with values ranging from 0 to n-1 for each row
+    index : Column or None
```
Suggest making this a column or a list of columns for the MultiIndex case

kkraus14 commented:
@rjzamora is this ready to be reviewed or still in progress?

rjzamora commented:
> @rjzamora is this ready to be reviewed or still in progress?
 
Sorry - not quite ready yet.  In order to handle categorical columns, I still need to add a workaround for the round-trip problem in scatter_to_frames

rjzamora commented:
@kkraus14 - I added the necessary categorical fix to `scatter_to_frames` (and improved testing).  So, this should be good for review (assuming CI passes).

kkraus14 commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -3929,7 +3939,7 @@ def scatter_by_map(self, map_index, map_size=None):
             # Append empty dataframes if map_size > len(tables)
             for i in range(map_size - len(tables)):
-                tables.append(self.iloc[[]])
+                tables.append(cudf.DataFrame(columns=self.columns))
```
Instead of returning an empty typed Index this will return a RangeIndex now, is that intended?

rjzamora replied:
Good point -  I was missing a check for empty-dataframe index types in `test_dataframe_scatter_by_map`, so I didn't notice this mistake.  I added a simple fix with test coverage.
